nalgebra-sparse/src/pattern.rs

```diff
@@ -184,6 +186,131 @@ impl SparsityPattern {
         })
     }
 
+    /// Try to construct a sparsity pattern from the given dimensions, major offsets
+    /// and minor indices.
+    ///
+    /// Returns an error if the data does not conform to the requirements.
+    pub fn from_offset_and_indices_unchecked(
+        major_dim: usize,
+        minor_dim: usize,
+        major_offsets: Vec<usize>,
+        minor_indices: Vec<usize>,
+    ) -> Result<Self, SparsityPatternFormatError> {
+        use SparsityPatternFormatError::*;
+
+        if major_offsets.len() != major_dim + 1 {
+            return Err(InvalidOffsetArrayLength);
+        }
+
+        // Check that the first and last offsets conform to the specification
+        {
+            let first_offset_ok = *major_offsets.first().unwrap() == 0;
+            let last_offset_ok = *major_offsets.last().unwrap() == minor_indices.len();
+            if !first_offset_ok || !last_offset_ok {
+                return Err(InvalidOffsetFirstLast);
+            }
+        }
+
+        Ok(Self {
+            major_offsets,
+            minor_indices,
+            minor_dim,
+        })
+    }
+
+    /// Validates cs data, sorts minor indices and values
+    pub fn validate_and_optionally_sort_cs_data<T>(
+        major_dim: usize,
+        minor_dim: usize,
+        major_offsets: &[usize],
+        minor_indices: &mut [usize],
+        values: Option<&mut [T]>,
+        sort: bool,
+    ) -> Result<(), SparsityPatternFormatError>
+    where
+        T: Scalar,
+    {
+        use SparsityPatternFormatError::*;
+
+        if major_offsets.len() != major_dim + 1 {
+            return Err(InvalidOffsetArrayLength);
+        }
+
+        // Check that the first and last offsets conform to the specification
+        {
+            let first_offset_ok = *major_offsets.first().unwrap() == 0;
+            let last_offset_ok = *major_offsets.last().unwrap() == minor_indices.len();
+            if !first_offset_ok || !last_offset_ok {
+                return Err(InvalidOffsetFirstLast);
+            }
+        }
+
+        let mut minor_index_permutation: Vec<usize> = (0..minor_indices.len()).collect();
+
+        // Test that each lane has strictly monotonically increasing minor indices, i.e.
+        // minor indices within a lane are sorted, unique. In addition, each minor index
+        // must be in bounds with respect to the minor dimension.
+        {
+            for lane_idx in 0..major_dim {
+                let range_start = major_offsets[lane_idx];
+                let range_end = major_offsets[lane_idx + 1];
+
+                // Test that major offsets are monotonically increasing
+                if range_start > range_end {
+                    return Err(NonmonotonicOffsets);
+                }
+
+                let minor_indices2 = &minor_indices[range_start..range_end];
```

Comment:
Maybe rather name the variable `minor_idx_in_lane` to make it clear that this is a slice of the minor indices in the current lane?

Also, the previous implementation didn't check the `range_start` and `range_end` (see discussion in #1019). Maybe we can incorporate the fix in this PR: you'd need to check that `range_start`/`range_end` are actually in bounds for `minor_indices`. You could then also port the test from #1019 to explicitly test for this case. What do you think?

Reply:
`minor_idx_in_lane` sounds better and I agree `range_start` and `range_end` should be checked.
 It looks like changes from https://github.com/dimforge/nalgebra/pull/1019 do not directly relate to the case of unsorted indices as we handle it in separate functions. I will check if I can add analogous tests for unsorted data constructor.